lib/rules/no-duplicate-selectors/index.mjs

```diff
@@ -50,124 +55,140 @@ const rule = (primary, secondaryOptions) => {
 		// from other rules that share the same parent and the same source.
 		const selectorContextLookup = nodeContextLookup();
 
-		root.walkRules((ruleNode) => {
-			if (isKeyframeRule(ruleNode)) {
-				return;
-			}
+		/**
+		 * @param {PostcssRule} node
+		 * @param {string} selector
+		 * @param {number} index
+		 * @param {number} previousDuplicateLine
+		 */
+		function complain(node, selector, index, previousDuplicateLine) {
+			report({
+				result,
+				ruleName,
+				node,
+				message: messages.rejected,
+				messageArgs: [selector, previousDuplicateLine],
+				index,
+				endIndex: index + selector.length,
+			});
+		}
 
-			const contextSelectorSet = selectorContextLookup.getContext(
-				ruleNode,
-				findAtRuleContext(ruleNode),
-			);
-			const resolvedSelectorList = [
-				...new Set(
-					ruleNode.selectors.flatMap((selector) => resolvedNestedSelector(selector, ruleNode)),
-				),
-			];
+		/**
+		 * Check selectors in the same rule:
+		 * `a, b, a {}` -> double "a"
+		 *
+		 * @param {PostcssRule} ruleNode
+		 */
+		function checkSelectorsInTheSameRule(ruleNode) {
+			const selectorsInList = new Map();
+			const selectors = parseSelector(getRuleSelector(ruleNode), result, ruleNode);
+
+			selectors?.each((selector) => {
+				const normalized = normalizeSelector(selector.clone()).toString();
+				const previousDuplicatePosition = selectorsInList.get(normalized);
+
+				if (!previousDuplicatePosition) {
+					selectorsInList.set(normalized, {
+						// line numbers start at 1, adding two line numbers together requires subtracting 1
+						line: (selector.source?.start?.line ?? 1) + (ruleNode.source?.start?.line ?? 1) - 1,
+					});
 
-			const normalizedSelectorList = resolvedSelectorList.map(normalize);
+					return;
+				}
 
-			// Sort the selectors list so that the order of the constituents
-			// doesn't matter
-			const sortedSelectorList = [...normalizedSelectorList].sort().join(',');
+				const index = getSelectorSourceIndex(selector);
+				const selectorStr = selector.toString().trim();
 
-			if (!ruleNode.source) throw new Error('The rule node must have a source');
+				complain(ruleNode, selectorStr, index, previousDuplicatePosition.line);
+			});
+		}
 
-			if (!ruleNode.source.start) throw new Error('The rule source must have a start position');
+		/**
+		 * Check selectors in different rules, but evaluate each selector list as a single item:
+		 * `a, b {} b, a {}` -> double "a, b"
+		 *
+		 * @param {PostcssRule} ruleNode
+		 * @param {Map<string, { line: number }>} contextSelectorSet
+		 */
+		function checkSelectorListInDifferentRules(ruleNode, contextSelectorSet) {
+			const combinedRoot = parser.root({
+				nodes: flattenNestedSelectorsForRule(ruleNode, result).flatMap(
+					({ resolvedSelectors }) => resolvedSelectors.nodes,
+				),
+				value: '',
+			});
 
-			const selectorLine = ruleNode.source.start.line;
+			const normalized = normalizeSelector(combinedRoot.clone()).toString();
```

Comment:
[ask] Just curious, is it worth avoiding `.clone()` like this?

```suggestion
			const normalized = normalizeSelector(combinedRoot).toString();
```

It seems that cloning may be unnecessary since `parser.root()` creates a new object. But, if the performance diff is too low, please leave it as-is.

Reply:
Good catch!
We can avoid it here and at line 151.

We still need it at line 88 because that variable is used in error reporting and should not be mutated.